specs/interop.md

```diff
@@ -0,0 +1,396 @@
+
+The exact definitions for these upgrade transactions are still to be defined.
+
+## Derivation Pipeline
+
+The derivation pipeline enforces invariants on safe blocks that include relaying transactions.
+
+- The relaying message MUST have a corresponding initiating message
+- The initiating message that corresponds to a relaying message MUST come from an allowed chain (dependency set)
+
+Blocks that contain transactions that relay cross domain messages to the destination chain where the
+initiating transaction does not exist MUST be considered invalid and MUST not be allowed by the
+derivation pipeline to be considered safe.
+
+There is no concept of replay protection within the protocol since there is no replay protection
+mechanism that fits well for all applications. Users MAY submit an arbitrary number of relaying
+messages per initiating message. This greatly simplifies the protocol and allows for
+applications to build application specific replay protection.
+
+### Depositing a Relaying Message
+
+Deposit transactions (force inclusion transactions) give censorship resistance to layer two networks.
+The derivation pipeline must gracefully handle the case in which a user uses a deposit transaction to
+relay a cross chain message. To not couple preconfirmation security to consensus, deposit transactions
+that relay cross chain messages MUST have an initiating message that is considered safe. This relaxes
+a strict synchrony assumption on the sequencer that it MUST have all unsafe blocks of destination chains
+as fast as possible to ensure that it is building correct blocks. It also prevents a class of attacks
+where the user can send a relaying message before the initiating message and trick the derivation pipeline
+into reorganizing the sequencer.
+
+### Safety
+
+The initiating messages for all relaying messages MUST be resolved as safe before an L2 block can transition from being unsafe to safe.
+Users MAY optimistically accept unsafe blocks without any verification of the relaying messages. They SHOULD optimistically verify
+the initiating messages exist in destination unsafe blocks to more quickly reorganize out invalid blocks.
+
+## State Transition Function
+
+After the full execution of a block, the set of `MessagePassed` events emitted from the `CrossL2Outbox` are collected and merklized.
+This commitment MUST be included as the block header's [extra data field](https://github.com/ethereum/execution-specs/blob/1fed0c0074f9d6aab3861057e1924411948dc50b/src/ethereum/frontier/fork_types.py#L115).
+The events are serialized with using [Simple Serialize](https://github.com/ethereum/consensus-specs/blob/dev/ssz/simple-serialize.md) aka SSZ.
+
+| Name          | Value |
+|---------------|-------------------------------------|
+| `MAX_LOG_DATA_SIZE` | `uint64(2**24)` or 16,777,216 |
+| `MAX_MESSAGES_PER_BLOCK` | `uint64(2**20)` or 1,048,576 |
+
+```python
+class CrossChainMessage(Container):
+    nonce: uint256
```

Comment:
nonce exists to ensure that only 1 log will ever exist per message. If it did not exist, then the same cross chain message hash could exist at multiple timestamps, which then the honest block builder would need to look at the latest timestamp. this adds edge cases so i recommend we keep nonce for simplicity here

Reply:
This no longer applies to the latest spec